### ISSUE TITLE
`[eslint] Maximum call stack size exceeded` error when using with `eslint-webpack-plugin`

### ISSUE THREAD

**Issue by 10hendersonm** [jest-community/eslint-plugin-jest#1724]
# Description

When using `eslint-plugin-jest` with [`eslint-webpack-plugin`](https://github.com/webpack-contrib/eslint-webpack-plugin?tab=readme-ov-file#eslint-webpack-plugin), a recursion error - `[eslint] Maximum call stack size exceeded` - occurs because...

1. `eslint-webpack-plugin` performs a `JSON.stringify` operation while evaluating the lint configuration objects
2. `eslint-plugin-jest` recurses `plugin` onto itself

# Implementation Details

Versions in use when this occurred:
- `eslint@9.19.0`
- `eslint-plugin-jest@28.11.0`
- `eslint-webpack-plugin@5.0.0`

```js
// webpack.config.js (invalid)
const ESLintWebpackPlugin = require('eslint-webpack-plugin')
const pluginJest = require('eslint-plugin-jest')

module.exports = function(webpackEnv) {
  // ...,
  plugins: [
    // ...,
    new ESLintWebpackPlugin({
      formatter: require.resolve('react-dev-utils/eslintFormatter'),
      eslintPath: require.resolve('eslint'),
      
      failOnError: true,

      extensions: ['js', 'mjs', 'jsx', 'ts', 'tsx'],
      cwd: '/some/path',
      context: '/some/path',

      configType: 'flat',
      baseConfig: [
        // ...,
        {
          plugins: { jest: pluginJest }
        }
      ],
    }),
  ],
}
```

`plugin` recurses in on itself (ex: `plugin.configs['flat/all'].plugins.jest === plugin`), meaning it can't be run through `JSON.stringify`.

https://github.com/jest-community/eslint-plugin-jest/blob/dc26b6224a2081dc7fda42b1dd2a884a5ae28d17/src/index.ts#L68-L110

This can be validated in a `node` REPL:

```sh
node

> JSON.stringify(require('eslint-plugin-jest'))
Uncaught TypeError: Converting circular structure to JSON
    --> starting at object with constructor 'Object'
    |     property 'configs' -> object with constructor 'Object'
    |     property 'flat/all' -> object with constructor 'Object'
    |     property 'plugins' -> object with constructor 'Object'
    --- property 'jest' closes the circle
    at JSON.stringify (<anonymous>)
```

This occurs with `eslint-webpack-plugin` whether or not the `baseConfig` is used (as opposed to the user's `eslint.config.js` file) as `eslint-webpack-plugin` evaluates the configuration it may be passing.

https://github.com/webpack-contrib/eslint-webpack-plugin/blob/a653827eb27875d2fc06054311a29afc285ba476/src/getESLint.js#L93-L119

# Workaround

My team and I managed to work around this by adding the plugin to our configuration in a relatively ugly way by plucking a specific object and cutting out the recursion:

```js
// webpack.config.js (workaround)
const ESLintWebpackPlugin = require('eslint-webpack-plugin')
const pluginJest = require('eslint-plugin-jest').configs['flat/all']
delete pluginJest.plugins.jest.configs

module.exports = function(webpackEnv) {
  // ...,
  plugins: [
    // ...,
    new ESLintWebpackPlugin({
      // ...,
      baseConfig: [
        // ...,
        {
          ...pluginJest,
          files: ['**/__tests__/**/*', '**/*.test.*'],
          rules: {
            // ...,
          }
        }
      ],
    }),
  ],
}
```

---

I'm not really sure who to say the bug is with here. On one hand, `eslint-plugin-webpack` isn't doing safe recursive JSON stringification, but on the other `eslint-plugin-jest`'s recursive `plugin` object is relatively dangerous.

**Comment by G-Rath** [comment]
> on the other eslint-plugin-jest's recursive plugin object is relatively dangerous.

This is how flat config works - plugins in configs are now the actual plugins, and since we're exporting a config we reference ourselves; our implementation mirrors what ESLint themselves tell you to do in their migration guide except they use `Object.assign` instead of the assignment operator: https://eslint.org/docs/latest/extend/plugin-migration-flat-config#migrating-configs-for-flat-config

**Comment by 10hendersonm** [comment]
Gotcha - thank you! I'm away from my desk but I'll re-open this on their end in the morning
